src/data/nat/pochhammer.lean

```diff
@@ -0,0 +1,293 @@
+-- they only use `n - 0 = n` and `n - m - k = n - (m + k)`.
+-- Without a `[monus R]` typeclass, we just prove them separately below in the `nat` namespace.
+variables [ring R]
+
+lemma falling_factorial_eq_mul_right {r : R} {n : ℕ} :
+  falling_factorial r (n + 1) = falling_factorial r n * (r - n) :=
+begin
+  induction n with n ih generalizing r,
+  { simp, },
+  { rw [falling_factorial, ih, falling_factorial, nat.succ_eq_add_one],
+    push_cast,
+    rw [mul_assoc, add_comm (n : R) 1, ←sub_sub], }
+end
+
+lemma falling_factorial_mul_falling_factorial {r : R} {n m : ℕ} :
+  falling_factorial r n * falling_factorial (r - n) m = falling_factorial r (n + m) :=
+begin
+  induction m with m ih,
+  { simp, },
+  { rw [falling_factorial_eq_mul_right, ←mul_assoc, ih, nat.add_succ,
+      falling_factorial_eq_mul_right],
+    push_cast,
+    rw [sub_sub], }
+end
+
+end
+
+namespace nat
+
+section
+variables [ring R]
+
+@[norm_cast]
+lemma falling_factorial_coe {r n : ℕ} :
+  ((falling_factorial r n : ℕ) : R) = falling_factorial (r : R) n :=
+begin
+  induction n with n ih generalizing r,
+  { simp, },
+  { simp only [falling_factorial, falling_factorial_eq_mul_left, nat.cast_mul],
+    rw [ih],
+    { by_cases w : r = 0,
+      { subst w, simp, },
+      { replace w : 0 < r := nat.pos_of_ne_zero w,
+        push_cast [w], }, }, },
+end
+
+/-- We already have this theorem for `r : R` with `[ring R]`,
+but need to prove it separately for `ℕ`.-/
+lemma falling_factorial_eq_mul_right {r n : ℕ} :
+  falling_factorial r (n + 1) = falling_factorial r n * (r - n) :=
+begin
+  -- We could prove this from the ring case by using the injectivity of `ℕ → ℤ`,
+  -- but it involves casing on `n ≤ r`, so it's easier to just redo it from scratch.
+  induction n with n ih generalizing r,
+  { simp, },
+  { rw [falling_factorial, ih, falling_factorial, succ_eq_add_one],
+    rw [mul_assoc, add_comm n 1, ←nat.sub_sub], }
+end
+
+@[simp]
+lemma falling_factorial_eq_factorial {n : ℕ} :
+  falling_factorial n n = n.factorial :=
+begin
+  induction n with n ih,
+  { simp, },
+  { simp [falling_factorial_eq_mul_left, ih], }
+end
+
+/-- We already have this theorem for `r : R` with `[ring R]`,
+but need to prove it separately for `ℕ`.-/
+lemma falling_factorial_mul_falling_factorial {r n m : ℕ} :
+  falling_factorial r n * falling_factorial (r - n) m = falling_factorial r (n + m) :=
+begin
+  induction m with m ih,
+  { simp, },
+  { rw [falling_factorial_eq_mul_right, ←mul_assoc, ih, add_succ,
+      falling_factorial_eq_mul_right, nat.sub_sub], }
+end
+
+lemma falling_factorial_ne_zero {n m : ℕ} (h : n ≤ m) :
+  falling_factorial m n ≠ 0 :=
+begin
+  intro w,
+  have := @falling_factorial_mul_falling_factorial m n (m-n),
+  rw [w, nat.add_sub_cancel' h, zero_mul, falling_factorial_eq_factorial] at this,
+  exact ne_of_lt m.factorial_pos this,
+end
+
+end
+
+end nat
+
+section
+variables [comm_ring R]
+
+lemma rising_factorial_eq_falling_factorial {r : R} {n : ℕ} :
```

Comment:
This holds in any `ring R`, since `(r - k) * (r - n) = r^2 - 2 * k * n * r + k * n = (r - n) * (r - k)`, right? Or is that too annoying to prove?

Reply:
Hmm... I've added a comment, but feel I should do this.

Thinking about it some more, I think this whole file could be improved with a different approach. I'm not particularly enthusiastic about implementing it right now, however.

We should introduce `pochhammer_polynomial (n : ℕ) : polynomial ℕ` (or probably even better, `polynomial S` for any semiring `S`, but for now that will only cloud the issue). This takes care automatically of the fact that `r-k` and `r-n` commute, because the polynomial ring itself is commutative.

Only then do we introduce `rising_factorial r n` as an evaluation of this polynomial at `r : R`.

Reply:
Ugh, I couldn't even get off the ground: https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/equation.20compiler.20complains.20about.20.60classical.2Echoice.60/near/228703412.